aggregation/txmetrics/aggregator.go

```diff
@@ -0,0 +1,364 @@
+			)
+		}
+	}
+}
+
+func (a *Aggregator) publish(ctx context.Context) error {
+	// We hold a.mu only long enough to swap the metrics. This will
+	// be blocked by metrics updates, which is OK, as we prefer not
+	// to block metrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+	a.active, a.inactive = a.inactive, a.active
+	a.mu.Unlock()
+
+	if len(a.inactive.buckets) == 0 {
+		a.logger.Debugf("no metrics to publish")
+		return nil
+	}
+
+	// TODO(axw) record either the aggregation interval in effect, or
+	// the specific time period (date_range) on the metrics documents.
+
+	now := time.Now()
+	metricsets := make([]transform.Transformable, 0, len(a.inactive.buckets))
+	for key, transactionMetrics := range a.inactive.buckets {
+		counts, values := transactionMetrics.buckets()
+		metricset := makeMetricset(key, now, counts, values)
+		metricsets = append(metricsets, &metricset)
+		delete(a.inactive.buckets, key)
+	}
+
+	a.logger.Debugf("publishing %d metricsets", len(metricsets))
+	return a.report(ctx, publish.PendingReq{
+		Transformables: metricsets,
+		Trace:          true,
+	})
+}
+
+// AggregateTransformables aggregates all transactions contained in
+// "in", returning the input with any metricsets requiring immediate
+// publication appended.
+//
+// This method is expected to be used immediately prior to publishing
+// the events, so that the metricsets requiring immediate publication
+// can be included in the same batch.
+func (a *Aggregator) AggregateTransformables(in []transform.Transformable) []transform.Transformable {
+	out := in
+	for _, tf := range in {
+		if tx, ok := tf.(*transaction.Event); ok {
+			if metricset := a.AggregateTransaction(tx); metricset != nil {
+				out = append(out, metricset)
+			}
+		}
+	}
+	return out
+}
+
+// AggregateTransaction aggregates transaction metrics.
+//
+// If the transaction cannot be aggregated due to the maximum number
+// of transaction groups being exceeded, then a *metricset.Metricset
+// will be returned which should be published immediately, along with
+// the transaction. Otherwise, the returned metricset will be nil.
+func (a *Aggregator) AggregateTransaction(tx *transaction.Event) *metricset.Metricset {
+	key := makeTransactionAggregationKey(tx)
+	duration := time.Duration(tx.Duration * float64(time.Millisecond))
+	if a.updateTransactionMetrics(key, duration) {
+		return nil
+	}
+	// Too many aggregation keys: could not update metrics, so immediately
+	// publish a single-value metric document.
+	//
+	// TODO(axw) log a warning with a rate-limit, increment a counter.
+	counts := []int64{1}
+	values := []float64{float64(durationMicros(duration))}
+	metricset := makeMetricset(key, time.Now(), counts, values)
+	return &metricset
+}
+
+func (a *Aggregator) updateTransactionMetrics(key transactionAggregationKey, duration time.Duration) bool {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	m := a.active
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
So far on publishing we only use a [read lock](https://github.com/elastic/apm-server/blob/master/publish/pub.go#L135) (except for when the server stops). With this addition a `Lock` is created for every transaction before being pushed to the `pendingRequests` channel, holding the lock while running the histogram related aggregation. Afaics this potentially creates a bottleneck for processing transaction events. 
This is additionally concerning since this logic runs while the APM agent still waits for a server response. 

We might want to consider something like pushing transactions into a temporary channel, do the aggregation when reading from the channel and only then pushing the transaction (+metricset) into the `pendingRequest` channel. Another option would be to do some refactoring allowing to do the metrics aggregation on transactions when reading from the `pendingRequest` channel before running the `transformation` on the event. 
I didn't look into any performance numbers, so might be good enough for now, but wondering if you measured this with hey-apm for concurrent requests?

Reply:
I haven't yet tested it with hey-apm, but will do so after I've rebased this on the other in-flight changes. My expectation is that it's good enough, but we'll see. ICYMI, there's a parallel benchmark in the txmetrics package:

```
BenchmarkAggregateTransaction            6073344               190 ns/op               0 B/op          0 allocs/op
BenchmarkAggregateTransaction-2          5032620               223 ns/op               0 B/op          0 allocs/op
BenchmarkAggregateTransaction-4          5218090               233 ns/op               0 B/op          0 allocs/op
BenchmarkAggregateTransaction-8          4581522               276 ns/op               0 B/op          0 allocs/op
BenchmarkAggregateTransaction-16         4438081               265 ns/op               0 B/op          0 allocs/op
```

I intend to revise this code later to minimise contention, following the approach taken for breakdown metrics in the Go agent: https://github.com/elastic/apm-agent-go/blob/839d3cc2699fdcc73b5524916519845724ec9d2d/breakdown.go#L225.

The idea would be to only take an exclusive lock for the whole map when recording a new group. We would still need to take an exclusive lock on the histogram/transaction group metrics though, unless we fork codahale/hdrhistogram and add a method that uses atomic increments.

Reply:
I prototyped the changes necessary to remove the exclusive lock, and forked hdrhistogram to add atomic update (turns out to be pretty straightforward).

```
BenchmarkAggregateTransaction            6636734               172 ns/op
BenchmarkAggregateTransaction-2          7426639               159 ns/op
BenchmarkAggregateTransaction-4          9878884               115 ns/op
BenchmarkAggregateTransaction-8         14216102                86.3 ns/op
BenchmarkAggregateTransaction-16        13393129                88.9 ns/op
```

There's a clear pattern there: previously adding goroutines decreased throughput (due to lock contention), now it improves.

I also ran both a hey-apm test, and a separate simple load test with multiple clients, and I saw no noticeable change with/without aggregation enabled. What do you think about going ahead with the current implementation, and then iterate on the implementation?

Reply:
SGTM - thanks for the investigation